sample-app-jsaddle/Material/TabBar.hs

```diff
@@ -0,0 +1,305 @@
+    ( Config, config
+    , setStacked
+    , setMinWidth
+    , setIndicatorSpansContent
+    , setAlign
+    , setAttributes
+    , tabBar
+    , Align(..)
+    ) where
+
+import qualified Data.Maybe as Maybe
+import qualified Data.Function
+import qualified Data.Void as Void
+import qualified Miso
+import qualified Miso.String
+import qualified Material.Tab.Internal as Tab
+
+(|>) = (Data.Function.&)
+(<|) = (Data.Function.$)
+(<<) = (Data.Function..)
+
+
+{-| Configuration of a tab bar
+-}
+data Config msg
+    = Config
+        { stacked :: Bool
+        , minWidth :: Bool
+        , indicatorSpansContent :: Bool
+        , additionalAttributes :: [Miso.Attribute msg]
+        , align :: Maybe Align
+        }
+
+
+{-| Default configuration of a tab bar
+-}
+config :: Config msg
+config =
+    Config
+        { stacked = False
+        , minWidth = False
+        , indicatorSpansContent = False
+        , align = Nothing
+        , additionalAttributes = []
+        }
+
+
+{-| Specify a tab bar's tabs to be stacked
+Stacked tabs display their icon below the their label.
+-}
+setStacked :: Bool -> Config msg -> Config msg
+setStacked stacked config_ =
+    config_ { stacked = stacked }
+
+
+{-| Specify whether a tab bar's tabs should be of minimum width
+Usually, a tab bar's tabs have a minimum with. Using this option, tabs are as
+narrow as possible.
+-}
+setMinWidth :: Bool -> Config msg -> Config msg
+setMinWidth minWidth config_ =
+    config_ { minWidth = minWidth }
+
+
+{-| Specify whether a tab bar's tab indicator spans its content
+Usually, a tab bar's tab indicator spans the entire tab. Use this option to
+make it span only it's label instead.
+-}
+setIndicatorSpansContent :: Bool -> Config msg -> Config msg
+setIndicatorSpansContent indicatorSpansContent config_ =
+    config_ { indicatorSpansContent = indicatorSpansContent }
+
+
+{-| Specify tab bar's alignment of tabs in case they overflow horizontally
+-}
+setAlign :: Maybe Align -> Config msg -> Config msg
+setAlign align config_ =
+    config_ { align = align }
+
+
+{-| Specify additional attribtues
+-}
+setAttributes :: [Miso.Attribute msg] -> Config msg -> Config msg
+setAttributes additionalAttributes config_ =
+    config_ { additionalAttributes = additionalAttributes }
+
+
+{-| Tab bar view function
+-}
+tabBar :: Config msg -> [Tab.Tab msg] -> Miso.View msg
+tabBar config_@Config { additionalAttributes=additionalAttributes, align=align } tabs =
+    Miso.nodeHtml "mdc-tab-bar"
+        (Maybe.mapMaybe id
+            [ rootCs
+            , tablistRoleAttr
+            , activeTabIndexProp tabs
+            ]
+            ++ additionalAttributes
+        )
+        [ tabScroller config_ align tabs ]
+
+
+rootCs :: Maybe (Miso.Attribute msg)
+rootCs =
+    Just (Miso.class_ "mdc-tab-bar")
+
+
+tablistRoleAttr :: Maybe (Miso.Attribute msg)
+tablistRoleAttr =
+    Just (Miso.textProp "role" "tablist")
+
+
+indexedMap :: (Int -> a -> b) -> [a] -> [b]
+indexedMap f xs =
```

Comment:
We should consider creating a module/library with the `Maybe`, `List`, `Map` APIs from elm (possibly every dependency of `material-components-web-elm`). It would avoid having to introduce inline helpers and the need of rewriting some of the code.

What do you think?

It doesn't need to be for this PR, but we can refactor the existing code to use this module/library.

Reply:
Those Elm style preludes exist.
However I'm a bit undecided currently on whether we should use them.
Same thing applies to using things like flow and lenses.
Elm code is incredibly nice to read so writing this in Elm styled Haskell is something I would find great.
On the other hand I want to write Haskell code that is easily understandable for Haskell beginners like myself.
What's you take on introducing these changes?